cmd/attest.go

```diff
@@ -0,0 +1,173 @@
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"strconv"
+
+	"cloud.google.com/go/compute/metadata"
+	"github.com/google/go-tpm-tools/client"
+	"github.com/google/go-tpm-tools/proto/attest"
+	"github.com/google/go-tpm/tpm2"
+	"github.com/spf13/cobra"
+	"google.golang.org/protobuf/proto"
+)
+
+var (
+	key      string
+	teeNonce []byte
+)
+
+var attestationKeys = map[string]map[tpm2.Algorithm]func(rw io.ReadWriter) (*client.Key, error){
+	"AK": {
+		tpm2.AlgRSA: client.AttestationKeyRSA,
+		tpm2.AlgECC: client.AttestationKeyECC,
+	},
+	"gceAK": {
+		tpm2.AlgRSA: client.GceAttestationKeyRSA,
+		tpm2.AlgECC: client.GceAttestationKeyECC,
+	},
+}
+
+var attestCmd = &cobra.Command{
+	Use:   "attest",
+	Short: "Create a remote attestation report",
+	Long: `Gather information for remote attestation.
+The Attestation report contains a quote on all available PCR banks, a way to validate 
+the quote, and a TCG Event Log (Linux only).
+Use --key to specify the type of attestation key. It can be gceAK for GCE attestation
+key or AK for a custom attestation key. By default it uses AK.
+--algo flag override the public key algorithm for attestation key. If not provided then
+by default rsa is used.
+--nonce attaches even length extra data to the report and guarantees a fresh quote.
+--teenonce attaches a 64 bytes extra data to the attestation report of hardware and 
+guarantees a fresh quote.
+--format flag specify the type of output file. It can be "binarypb" or "textproto".
+By default it uses binarypb.
+--output flag provides file name to store attestation report in the format specified
+by --format flag.
+`,
+	Args: cobra.NoArgs,
+	RunE: func(cmd *cobra.Command, args []string) error {
+
+		rwc, err := openTpm()
+		if err != nil {
+			return err
+		}
+		defer rwc.Close()
+
+		var attestationKey *client.Key
+		algoToCreateAK, ok := attestationKeys[key]
+		if !ok {
+			return fmt.Errorf("key should be either AK or gceAK")
+		}
+		createFunc := algoToCreateAK[keyAlgo]
+		attestationKey, err = createFunc(rwc)
+		if err != nil {
+			return fmt.Errorf("failed to create attestation key: %v", err)
+		}
+		defer attestationKey.Close()
+
+		attestOpts := client.AttestOpts{}
+		attestOpts.Nonce = nonce
+		if len(teeNonce) != 0 {
+			attestOpts.TEENonce = teeNonce
+			attestOpts.TEEDevice, err = client.CreateSevSnpDevice()
```

Comment:
getTeeAttestationReport will choose the device if TEEDevice is not set, so I don't think you need to make that default decision here.

Reply:
In `getTeeAttestationReport` , there is a check 
```
// TEEDevice can't be nil while TEENonce is non-nil
	if opts.TEENonce != nil {
		return fmt.Errorf("got non-nil TEENonce when TEEDevice is nil: %v", opts.TEENonce)
	}
```
And if the user provides a TEENonce flag , then it always fails that check because TEEDevice is nil.

Reply:
Perhaps we should just default to using the nonce as the teenonce (and then remove the option). That's what client.Attest does if it successfully creates an SNP device but doesn't get a TEENonce. 

Or, we can add a new flag requesting the hardware type (`hwattestation`?) as part of requesting hardware attestation. And teenonce would fail without that flag.

Either way, we need to be forwards compatible with new technologies like TDX.

Reply:
It was my intention for teenonce to take the value of nonce if teenonce is not populated.